crates/rspack_binding_values/src/filename.rs

```diff
@@ -0,0 +1,109 @@
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use napi::{
+  bindgen_prelude::{FromNapiValue, ToNapiValue, ValidateNapiValue},
+  Either, JsFunction,
+};
+use rspack_core::{AssetInfo, LocalFilenameFn, PathData};
+use rspack_core::{Filename, FilenameFn};
+use rspack_napi::{threadsafe_function::ThreadSafeFunctionWithRef, ByRef, NapiResultExt};
+use serde::Deserialize;
+
+use crate::{JsAssetInfo, PathData as JsPathData};
+
+/// A js filename value. Either a string or a function
+///
+/// The function type is generic. By default the function type is thread-safe and implements `Hash` and `Eq` by reference.
+#[derive(Debug)]
+pub struct JsFilename<
+  F = ByRef<ThreadSafeFunctionWithRef<(JsPathData, Option<JsAssetInfo>), String>>,
```

Comment:
I might not recommend you to create an additional isomorphic tsfn. By default, `JsFunction` and `ThreadsafeFunction` are two different kind of stuff. Trying to merge these two together would cause its state and types being implicit and cause potential deadlocks.